jerry-main/main-unix-snapshot.c

```diff
@@ -0,0 +1,411 @@
+                        lit_mem_to_snapshot_id_map_entry_t *lit_map_p) /**< literal map */
+{
+  JERRY_ASSERT (buffer_end_p > buffer_p);
+
+  do
+  {
+    ecma_compiled_code_t *bytecode_p = (ecma_compiled_code_t *) buffer_p;
+    uint32_t code_size = ((uint32_t) bytecode_p->size) << JMEM_ALIGNMENT_LOG;
+
+    if (bytecode_p->status_flags & CBC_CODE_FLAGS_FUNCTION)
+    {
+      jmem_cpointer_t *literal_start_p;
+      uint32_t const_literal_end;
+
+      if (bytecode_p->status_flags & CBC_CODE_FLAGS_UINT16_ARGUMENTS)
+      {
+        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint16_arguments_t));
+
+        cbc_uint16_arguments_t *args_p = (cbc_uint16_arguments_t *) buffer_p;
+        const_literal_end = args_p->const_literal_end;
+      }
+      else
+      {
+        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint8_arguments_t));
+
+        cbc_uint8_arguments_t *args_p = (cbc_uint8_arguments_t *) buffer_p;
+        const_literal_end = args_p->const_literal_end;
+      }
+
+      for (uint32_t i = 0; i < const_literal_end; i++)
+      {
+        if (literal_start_p[i] != JMEM_CP_NULL)
+        {
+          lit_mem_to_snapshot_id_map_entry_t *current_p = lit_map_p;
+
+          while (current_p->literal_id != literal_start_p[i])
+          {
+            current_p++;
+          }
+
+          literal_start_p[i] = current_p->literal_offset;
+        }
+      }
+    }
+
+    buffer_p += code_size;
+  }
+  while (buffer_p < buffer_end_p);
+} /* update_literal_offsets */
+
+/**
+ * Print help.
+ */
+static void
+print_help (char *name) /**< program name */
+{
+  printf ("Usage: %s [COMMAND] [OUT_FILE] [INPUT_FILEs]...\n"
+          "\n"
+          "Command:\n"
+          "  merge : merge input files into the output file\n"
+          "  help : print help\n"
+          "\n",
+          name);
+} /* print_help */
+
+/**
+ * Main function.
+ *
+ * @return error code (0 - no error)
+ */
+int
+main (int argc, /**< number of arguments */
+      char **argv) /**< argument list */
+{
+  if (argc <= 1 || (argc == 2 && (!strcmp ("help", argv[1]))))
+  {
+    print_help (argv[0]);
+    return JERRY_STANDALONE_EXIT_CODE_OK;
+  }
+
+  if (strcmp ("merge", argv[1]))
+  {
+    printf ("Unknown command: %s\n", argv[1]);
+    print_help (argv[0]);
+    return JERRY_STANDALONE_EXIT_CODE_OK;
+  }
+
+  if (argc < 5)
+  {
+    printf ("At least an output and two input files must be passed\n");
+    return JERRY_STANDALONE_EXIT_CODE_FAIL;
+  }
+
+  jerry_init (JERRY_INIT_EMPTY);
+
+  snapshot_context_t context;
+  memset (&context, 0, sizeof (snapshot_context_t));
+  context.input_pos_p = input_buffer;
+
+  for (int i = 3; i < argc; i++)
+  {
+    if (!read_file (&context, argv[i]))
+    {
+      return JERRY_STANDALONE_EXIT_CODE_FAIL;
+    }
+  }
+
+  snapshot_data_t *snapshot_p = context.first_snapshot_p;
+  uint32_t number_of_funcs = 0;
+  size_t functions_size = sizeof (jerry_snapshot_header_t);
+
+  while (snapshot_p != NULL)
```

Comment:
I've taken a closer look at the code and at the possibility of separating reusable logic from driver code (i.e., from code in a command line tool). I think it is not that hard to split them, actually, and it would be possible to come up with a single function call on library side that fits quite well into the existing public jerry-core snapshot API (no matter whether considered or marked as experimental or not).

@zherczeg I'd be glad if you'd consider the idea manifested in the following snippet. (There may be some loose ends but hopefully not too many. Actually, this is very much the code you wrote, just reorganized. Variable names, control structures, etc. kept as much as possible. The main change is that there is no need for snapshot contexts and linked lists, because VLAs help on the tool side -- and the library can work then with traditional arrays.)

```c
/* Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string.h>

#include "byte-code.h"
#include "ecma-literal-storage.h"
#include "jerry-snapshot.h"
#include "jerryscript.h"
#include "jerryscript-port.h"
#include "jmem.h"

//// CORE

/**
 * Collect all literals from a snapshot file.
 */
static void
scan_snapshot_functions (const uint8_t *buffer_p, /**< snapshot buffer start */
                         const uint8_t *buffer_end_p, /**< snapshot buffer end */
                         const uint8_t *literal_base_p, /**< start of literal data */
                         const uint8_t *number_base_p) /**< start of number data */
{
  JERRY_ASSERT (buffer_end_p > buffer_p);

  do
  {
    ecma_compiled_code_t *bytecode_p = (ecma_compiled_code_t *) buffer_p;
    uint32_t code_size = ((uint32_t) bytecode_p->size) << JMEM_ALIGNMENT_LOG;

    if (bytecode_p->status_flags & CBC_CODE_FLAGS_FUNCTION)
    {
      jmem_cpointer_t *literal_start_p;
      uint32_t const_literal_end;

      if (bytecode_p->status_flags & CBC_CODE_FLAGS_UINT16_ARGUMENTS)
      {
        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint16_arguments_t));

        cbc_uint16_arguments_t *args_p = (cbc_uint16_arguments_t *) buffer_p;
        const_literal_end = args_p->const_literal_end;
      }
      else
      {
        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint8_arguments_t));

        cbc_uint8_arguments_t *args_p = (cbc_uint8_arguments_t *) buffer_p;
        const_literal_end = args_p->const_literal_end;
      }

      for (uint32_t i = 0; i < const_literal_end; i++)
      {
        if (literal_start_p[i] != JMEM_CP_NULL)
        {
          literal_start_p[i] = ecma_snapshot_get_literal (literal_base_p, number_base_p, literal_start_p[i]);
        }
      }
    }

    buffer_p += code_size;
  }
  while (buffer_p < buffer_end_p);
} /* scan_snapshot_functions */

/**
 * Update all literal offsets in a snapshot data.
 */
static void
update_literal_offsets (uint8_t *buffer_p, /**< snapshot buffer start */
                        const uint8_t *buffer_end_p, /**< snapshot buffer start */
                        lit_mem_to_snapshot_id_map_entry_t *lit_map_p) /**< literal map */
{
  JERRY_ASSERT (buffer_end_p > buffer_p);

  do
  {
    ecma_compiled_code_t *bytecode_p = (ecma_compiled_code_t *) buffer_p;
    uint32_t code_size = ((uint32_t) bytecode_p->size) << JMEM_ALIGNMENT_LOG;

    if (bytecode_p->status_flags & CBC_CODE_FLAGS_FUNCTION)
    {
      jmem_cpointer_t *literal_start_p;
      uint32_t const_literal_end;

      if (bytecode_p->status_flags & CBC_CODE_FLAGS_UINT16_ARGUMENTS)
      {
        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint16_arguments_t));

        cbc_uint16_arguments_t *args_p = (cbc_uint16_arguments_t *) buffer_p;
        const_literal_end = args_p->const_literal_end;
      }
      else
      {
        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint8_arguments_t));

        cbc_uint8_arguments_t *args_p = (cbc_uint8_arguments_t *) buffer_p;
        const_literal_end = args_p->const_literal_end;
      }

      for (uint32_t i = 0; i < const_literal_end; i++)
      {
        if (literal_start_p[i] != JMEM_CP_NULL)
        {
          lit_mem_to_snapshot_id_map_entry_t *current_p = lit_map_p;

          while (current_p->literal_id != literal_start_p[i])
          {
            current_p++;
          }

          literal_start_p[i] = current_p->literal_offset;
        }
      }
    }

    buffer_p += code_size;
  }
  while (buffer_p < buffer_end_p);
} /* update_literal_offsets */

/**
 * Merge multiple snapshots into one.
 *
 * @return length of the merged snapshot - if successful,
 *         0 - otherwise.
 */
size_t
jerry_merge_snapshots (uint8_t **in_buffer_pp, /**< array of (pointers to start of) input buffers */
                       size_t *in_buffer_size_p /**< array of input buffer sizes */
                       size_t in_buffer_cnt, /**< number of input buffers */
                       uint8_t *out_buffer_p, /**< (pointer to start of) output buffer */
                       size_t out_buffer_size) /**< size of output buffer */
{
  JERRY_UNUSED (in_buffer_size_p); /* it could be useful for memory safety but is unused right now */

  uint32_t number_of_funcs = 0;
  size_t functions_size = sizeof (jerry_snapshot_header_t);

  for (size_t i = 0; i < in_buffer_cnt; i++)
  {
    jerry_snapshot_header_t *header_p = (jerry_snapshot_header_t *) in_buffer_pp[i];

    uint32_t start_offset = header_p->func_offsets[0] & ~JERRY_SNAPSHOT_EVAL_CONTEXT;
    const uint8_t *data_p = (const uint8_t *) header_p;
    const uint8_t *literal_base_p;
    const uint8_t *number_base_p;

    literal_base_p = ecma_snapshot_get_literals_base ((uint32_t *) (data_p + header_p->lit_table_offset),
                                                      &number_base_p);

    JERRY_ASSERT (header_p->number_of_funcs > 0);

    number_of_funcs += header_p->number_of_funcs;
    functions_size += header_p->lit_table_offset - start_offset;

    scan_snapshot_functions (data_p + start_offset,
                             data_p + header_p->lit_table_offset,
                             literal_base_p,
                             number_base_p);
  }

  JERRY_ASSERT (number_of_funcs > 0);

  functions_size += JERRY_ALIGNUP ((number_of_funcs - 1) * sizeof (uint32_t), JMEM_ALIGNMENT);

  if (functions_size >= out_buffer_size)
  {
    return 0; /* Output buffer is too small */
  }

  jerry_snapshot_header_t *header_p = (jerry_snapshot_header_t *) out_buffer_p;

  header_p->version = JERRY_SNAPSHOT_VERSION;
  header_p->lit_table_offset = (uint32_t) functions_size;
  header_p->number_of_funcs = number_of_funcs;

  lit_mem_to_snapshot_id_map_entry_t *lit_map_p;
  uint32_t literals_num;

  if (!ecma_save_literals_for_snapshot ((uint32_t *) out_buffer_p,
                                        out_buffer_size,
                                        &functions_size,
                                        &lit_map_p,
                                        &literals_num))
  {
    return 0; /* Not enough space for literals */
  }

  uint32_t *func_offset_p = header_p->func_offsets;
  uint8_t *dst_p = out_buffer_p + sizeof (jerry_snapshot_header_t);
  dst_p += JERRY_ALIGNUP ((number_of_funcs - 1) * sizeof (uint32_t), JMEM_ALIGNMENT);

  for (size_t i = 0; i < in_buffer_cnt; i++)
  {
    jerry_snapshot_header_t *current_header_p = (jerry_snapshot_header_t *) in_buffer_pp[i];

    uint32_t start_offset = current_header_p->func_offsets[0] & ~JERRY_SNAPSHOT_EVAL_CONTEXT;

    memcpy (dst_p,
            ((const uint8_t *) current_header_p) + start_offset,
            current_header_p->lit_table_offset - start_offset);

    update_literal_offsets (dst_p,
                            dst_p + current_header_p->lit_table_offset - start_offset,
                            lit_map_p);

    uint32_t current_offset = (uint32_t) (dst_p - out_buffer_p) - start_offset;

    for (uint32_t i = 0; i < current_header_p->number_of_funcs; i++)
    {
      /* Updating offset without changing any flags. */
      *func_offset_p++ = current_header_p->func_offsets[i] + current_offset;
    }

    dst_p += current_header_p->lit_table_offset - start_offset;
  }

  JERRY_ASSERT (dst_p - out_buffer_p == header_p->lit_table_offset);

  jmem_heap_free_block (lit_map_p, literals_num * sizeof (lit_mem_to_snapshot_id_map_entry_t));

  return functions_size;
} /* jerry_merge_snapshots */

//// TOOL

/**
 * Maximum size for loaded snapshots
 */
#define JERRY_BUFFER_SIZE (1048576)

/**
 * Standalone Jerry exit codes
 */
#define JERRY_STANDALONE_EXIT_CODE_OK   (0)
#define JERRY_STANDALONE_EXIT_CODE_FAIL (1)

static uint8_t input_buffer[ JERRY_BUFFER_SIZE ];
static uint8_t output_buffer[ JERRY_BUFFER_SIZE ];

/**
 * Loading a single file into the memory.
 *
 * @return length of the read file - loading is successful
 *         0 - otherwise
 */
static size_t
read_file (uint8_t *input_pos_p, /**< position to read file into */
           const char *file_name) /**< file name */
{
  FILE *file = fopen (file_name, "r");
  if (file == NULL)
  {
    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: failed to open file: %s\n", file_name);
    return 0;
  }

  size_t max_size = (size_t) (input_buffer + JERRY_BUFFER_SIZE - input_pos_p);

  size_t bytes_read = fread (input_pos_p, 1u, max_size, file);
  fclose (file);

  if (!bytes_read)
  {
    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: failed to read file: %s\n", file_name);
    return 0;
  }

  if (bytes_read >= max_size)
  {
    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: file too large: %s\n", file_name);
    return 0;
  }

  if (bytes_read < sizeof (jerry_snapshot_header_t))
  {
    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: invalid snapshot file: %s\n", file_name);
    return 0;
  }

  jerry_snapshot_header_t *header_p = (jerry_snapshot_header_t *) input_pos_p;

  if (header_p->version != JERRY_SNAPSHOT_VERSION)
  {
    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: invalid snapshot version: %s\n", file_name);
    return 0;
  }

  printf ("Input file '%s' loaded. Contains %d function(s).\n", file_name, header_p->number_of_funcs);

  return JERRY_ALIGNUP (bytes_read, JMEM_ALIGNMENT);
} /* read_file */

/**
 * Print help.
 */
static void
print_help (char *name) /**< program name */
{
  printf ("Usage: %s [COMMAND] [ARGUMENTS]\n"
          "\n"
          "Command list:\n"
          "  merge [OUTPUT_FILE] [INPUT_FILEs]...\n"
          "      Merge input files into output file.\n"
          "  help\n"
          "      Prints this help.\n"
          "\n",
          name);
} /* print_help */

/**
 * Main function.
 *
 * @return error code (0 - no error)
 */
int
main (int argc, /**< number of arguments */
      char **argv) /**< argument list */
{
  if (argc <= 1 || (argc == 2 && (!strcmp ("help", argv[1]))))
  {
    print_help (argv[0]);
    return JERRY_STANDALONE_EXIT_CODE_OK;
  }

  if (strcmp ("merge", argv[1]))
  {
    printf ("Unknown command: %s\n", argv[1]);
    print_help (argv[0]);
    return JERRY_STANDALONE_EXIT_CODE_OK;
  }

  if (argc < 5)
  {
    printf ("At least an output and two input files must be passed\n");
    return JERRY_STANDALONE_EXIT_CODE_FAIL;
  }

  jerry_init (JERRY_INIT_EMPTY);

  size_t input_count = argc - 3;
  uint8_t *input_buffers[input_count];
  size_t input_sizes[input_count];
  uint8_t *input_pos_p = input_buffer;

  for (int i = 0; i < input_count; i++)
  {
    size_t input_size = read_file (input_pos_p, argv[3 + i]);
    if (input_size == 0)
    {
      jerry_cleanup ();
      return JERRY_STANDALONE_EXIT_CODE_FAIL;
    }

    input_buffers[i] = input_pos_p;
    input_sizes[i] = input_size;
    input_pos_p += input_size;
  }

  size_t output_size = jerry_merge_snapshots (input_buffers, input_sizes, input_count,
                                              output_buffer, JERRY_BUFFER_SIZE);
  if (output_size == 0)
  {
    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Cannot merge snapshots\n");
    jerry_cleanup ();
    return JERRY_STANDALONE_EXIT_CODE_FAIL;
  }

  FILE *file = fopen (argv[2], "w");
  if (file == NULL)
  {
    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Cannot open file: '%s'\n", argv[2]);
    jerry_cleanup ();
    return JERRY_STANDALONE_EXIT_CODE_FAIL;
  }

  fwrite (output_buffer, 1u, output_size, file);
  fclose (file);

  jerry_cleanup ();
  return JERRY_STANDALONE_EXIT_CODE_OK;
} /* main */
```